src/mlpack/methods/ann/layer/grouped_convolution.hpp

```diff
@@ -0,0 +1,427 @@
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_GROUPED_CONVOLUTION_HPP
+#define MLPACK_METHODS_ANN_LAYER_GROUPED_CONVOLUTION_HPP
+
+#include <mlpack/prereqs.hpp>
+
+#include <mlpack/methods/ann/convolution_rules/border_modes.hpp>
+#include <mlpack/methods/ann/convolution_rules/naive_convolution.hpp>
+#include <mlpack/methods/ann/convolution_rules/fft_convolution.hpp>
+#include <mlpack/methods/ann/convolution_rules/svd_convolution.hpp>
+#include <mlpack/core/util/to_lower.hpp>
+
+#include "layer.hpp"
+#include "padding.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * Implementation of the Grouped Convolution class. The Grouped Convolution 
+ * class represents a single layer of a neural network.
+ * Example usage:
+ *
+ * Suppose we want to pass a matrix M (2744x100) to a `Convolution` layer;
+ * in this example, `M` was obtained from "flattening" 100 images (or Mel
+ * cepstral coefficients, if we talk about speech, or whatever you like) of
+ * dimension 196x14. In other words, the first 196 columns of each row of M
+ * will be made of the 196 columns of the first row of each of the 100 images
+ * (or Mel cepstral coefficients). Then the next 295 columns of M (196 - 393)
+ * will be made of the 196 columns of the second row of the 100 images (or Mel
+ * cepstral coefficients), etc.  Given that the size of our 2-D input images is
+ * 196x14, the parameters for our `Convolution` layer will be something like
+ * this:
+ *
+ * ```
+ * GroupedConvolution<> c(1, // Number of input activation maps.
+ *                        14, // Number of output activation maps.
+ *                        3, // Filter width.
+ *                        3, // Filter height.
+ *                        1, // Number of groups.
+ *                        1, // Stride along width.
+ *                        1, // Stride along height.
+ *                        0, // Padding width.
+ *                        0, // Padding height.
+ *                        196, // Input width.
+ *                        14); // Input height.
+ * ```
+ *
+ * This `Convolution<>` layer will treat each column of the input matrix `M` as
+ * a 2-D image (or object) of the original 196x14 size, using this as the input
+ * for the 14 filters of this example.
+ * 
+ * Groups controls the connections between inputs and outputs. 
+ * inMaps and outMaps must both be divisible by groups. 
+ * For example,
+ *    At groups=1, all inputs are convolved to all outputs.
+ *    At groups=2, the operation becomes equivalent to having two conv layers 
+ *      side by side, each seeing half the input channels and producing half 
+ *      the output channels, and both subsequently concatenated.
+ *    At groups= inMaps, each input channel is convolved with its own set of 
+ *      filters (of size \frac{\text{out\_channels}}{\text{in\_channels}}).
```

Comment:
https://pytorch.org/docs/stable/generated/torch.nn.Conv2d.html

Reply:
I don't think linking to the PyTorch documentation for something that's not specifically a grouped convolution is a great idea.  There are differences between PyTorch and mlpack, and because of that it's not a great idea to make users believe that mlpack is a reimplementation of PyTorch.  Also, the idea of referencing a paper is to get the mathematical idea of what we are implementing across.  For instance, I think that [this paper](https://openaccess.thecvf.com/content_cvpr_2018/papers/Huang_CondenseNet_An_Efficient_CVPR_2018_paper.pdf) has a nice description of grouped convolutions, in Section 2.3 and Figure 2.  Maybe you can add a reference to this paper to the documentation, in BibTeX format like we do with the rest of the algorithms?  Or another paper is fine too if you can find a better one.  (The original AlexNet paper doesn't specifically describe the technique, even if it was the first to use grouped convolutions.)